backend/audit/cross_validation/submission_progress_check.py

```diff
@@ -0,0 +1,131 @@
+    """
+    Because this function was initially created in a view and not as a
+    cross-validation function, it needs the crossval argument to distinguish between
+    running in a cross-validation context and running in the context of needing to
+    return progress information to the submission progress view.
+
+    crossval defaults to True because we don't want to have to change all the calls to
+    the validation functions to include this argument.
+
+    Given a SingleAuditChecklist instance and a SingleAuditReportFile instance,
+    return information about submission progress.
+
+    Returns this shape:
+
+        {
+            "complete": [bool],
+            "single_audit_report": [progress_dict],
+            "additional_ueis": [progress_dict],
+            ...
+            "general_information": [progress_dict],
+        }
+
+    Where each of the sections is represented at the top level, along with
+    single_audit_report, and [progress_dict] is:
+
+        {
+            "display": "hidden"/"incomplete"/"complete",
+            "completed": [bool],
+            "completed_by": [email],
+            "completed_date": [date],
+        }
+    """
+    sections = sac["sf_sac_sections"]
+    # TODO: remove these once Notes to SEFA and tribal data consent are implemented
+    del sections["notes_to_sefa"]
+    del sections["tribal_data_consent"]
+    result = {k: None for k in sections}  # type: ignore
+    progress = {
+        "display": None,
+        "completed": None,
+        "completed_by": None,
+        "completed_date": None,
+    }
+
+    cond_keys = _conditional_keys_progress_check(sections)
+    for ckey, cvalue in cond_keys.items():
+        result[ckey] = progress | cvalue
+
+    mandatory_keys = _mandatory_keys_progress_check(sections, cond_keys)
+    for mkey, mvalue in mandatory_keys.items():
+        result[mkey] = progress | mvalue
+
+    sar_progress = {
+        "display": "complete" if bool(sar) else "incomplete",
+        "completed": bool(sar),
+    }
+
+    result["single_audit_report"] = progress | sar_progress  # type: ignore
+
+    complete = False
+
+    def cond_pass(cond_key):
+        passing = ("hidden", "complete")
+        return result.get(cond_key, {}).get("display") in passing
+
+    error_keys = (
+        list(mandatory_keys.keys()) + list(cond_keys.keys()) + ["single_audit_report"]
+    )
+
+    # Need this to return useful errors in cross-validation:
+    incomplete_sections = [
+        k for k in error_keys if result[k].get("display") == "incomplete"
+    ]
+
+    if all(bool(sections[k]) for k in mandatory_keys):
+        if all(cond_pass(j) for j in cond_keys):
+            complete = True
+
+    result["complete"] = complete  # type: ignore
+
+    if not crossval:
+        return result  # return the cross-validation shape
+
+    if complete:
+        return []
+
+    return [
+        {
+            "error": f"The following sections are incomplete: {', '.join(incomplete_sections)}."
```

Comment:
given that we're statically returning a single `dict` item in this `array`, would it make sense for `result` to always be a `dict` which conditionally includes a top-level `error` key? In the `not crossval` case, the `error` (if present) could just be ignored by the caller